### PR TITLE
Add `BoundKind` in `visit_param_bounds` to check questions in bounds

### PR DESCRIPTION
**Pull request by SparrowLii**
From the FIXME in the impl of `AstValidator`. Better bound checks by adding `BoundCtxt` type parameter to `visit_param_bound`

cc @ecstatic-morse

### CODE REVIEW

File: compiler/rustc_ast/src/visit.rs [hunk 1]
```diff
@@ -32,6 +32,13 @@ pub enum FnCtxt {
     Assoc(AssocCtxt),
 }
 
+#[derive(Copy, Clone, Debug)]
+pub enum BoundCtxt {
+    Normal,
```
Comment (compiler-errors) [inline compiler/rustc_ast/src/visit.rs:37]: Can you leave some comments for what these mean in AST/HIR? Like "normal" is `<T: Bound>` and `where T: Bound` or something?

Comment (compiler-errors) [inline compiler/rustc_ast/src/visit.rs:37]: Not a huge fan of "Normal"... can you think of a better name? :)

File: compiler/rustc_ast/src/visit.rs [hunk 2]
```diff
@@ -32,6 +32,13 @@ pub enum FnCtxt {
     Assoc(AssocCtxt),
 }
 
+#[derive(Copy, Clone, Debug)]
+pub enum BoundCtxt {
```
Comment (compiler-errors) [inline compiler/rustc_ast/src/visit.rs:36]: Maybe `BoundKind`?

File: compiler/rustc_ast/src/visit.rs [hunk 3]
```diff
@@ -416,8 +423,11 @@ pub fn walk_ty<'a, V: Visitor<'a>>(visitor: &mut V, typ: &'a Ty) {
             visitor.visit_ty(ty);
             visitor.visit_anon_const(length)
         }
-        TyKind::TraitObject(ref bounds, ..) | TyKind::ImplTrait(_, ref bounds) => {
-            walk_list!(visitor, visit_param_bound, bounds);
+        TyKind::TraitObject(ref bounds, ..) => {
+            walk_list!(visitor, visit_param_bound, bounds, BoundCtxt::TraitObject);
+        }
+        TyKind::ImplTrait(_, ref bounds) => {
+            walk_list!(visitor, visit_param_bound, bounds, BoundCtxt::Normal);
```
Comment (compiler-errors) [inline compiler/rustc_ast/src/visit.rs:430]: Are these `::Normal`? They might need to be their own kind...

**Review by compiler-errors — CHANGES_REQUESTED** [review]
Thanks for the contribution! This is certainly better looking than the previous way. I have a few nits, then I can take a closer look and verify this works as before.

**Comment by SparrowLii** [comment]
Thanks for reviewing! I renamed `BoundCtxt` to `BoundKind` and added comment on the variants.

File: compiler/rustc_ast/src/visit.rs (hunk 3 above)
Comment (SparrowLii) [reply compiler/rustc_ast/src/visit.rs]: Yea, while there is no special check for it in the ast validity check, I agree that it is reasonable to differentiate it

**Pull request merged**